src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -773,7 +773,9 @@ public override ForEachStatementInfo GetForEachStatementInfo(ForEachStatementSyn
 
         public override ForEachStatementInfo GetForEachStatementInfo(CommonForEachStatementSyntax node)
         {
-            BoundForEachStatement boundForEach = (BoundForEachStatement)GetUpperBoundNode(node);
+            // Declaration expressions can result in having a BoundBlock associated with a given foreach statement,
+            // in addition to a BoundForEachStatement
+            BoundForEachStatement boundForEach = GetBoundNodes(node).OfType<BoundForEachStatement>().FirstOrDefault();
```

Comment:
Note to self, so I don't forget: 
What we discussed last was introducing a "raw" cache in the semantic model, mapping syntax to bound node. It would contain the nodes that result from binding (if you bind this syntax, this is the bound node you get back), without any filtering.
Most code paths in the semantic model would continue to rely on the existing cache, which filters with `ShouldAddNode`.